github-link.php

```diff
@@ -1,180 +1,168 @@
+Description:       Displays GitHub icon link on the Plugins page for a given <code>GitHub Plugin URI</code> plugin header.
 License:           The MIT License (MIT)
 Author:            Viktor Szépe
 Domain Path:       /languages
 Text Domain:       github-link
 GitHub Plugin URI: https://github.com/szepeviktor/github-link
 */
 
-if ( ! function_exists( 'add_filter' ) ) {
-    error_log( 'Malicious traffic detected: github_link_direct_access '
-        . addslashes( $_SERVER['REQUEST_URI'] )
-    );
-    ob_get_level() && ob_end_clean();
-    header( 'Status: 403 Forbidden' );
-    header( 'HTTP/1.0 403 Forbidden' );
-    exit();
+/**
+ * Exit if called directly.
+ */
+if ( ! defined( 'WPINC' ) ) {
+	die;
 }
 
-load_plugin_textdomain( 'github-link', false, dirname( __FILE__ ) . '/languages' );
-
-add_filter( "extra_plugin_headers", "GHL_extra_headers" );
-add_filter( "plugin_action_links", "GHL_plugin_link", 10, 4 );
-add_filter( "network_admin_plugin_action_links", "GHL_plugin_link", 10, 4 );
-add_action( "admin_head-plugins.php", "GHL_wp49_css_fix" );
-
-function GHL_extra_headers( $extra_headers ) {
-
-    // Keys will get lost.
-    return array_merge( $extra_headers, array(
-        "GitHubURI" => "GitHub Plugin URI",
-        "GitHubBranch" => "GitHub Branch",
-        "GitHubToken" => "GitHub Access Token",
-        "GitLabURI" => "GitLab Plugin URI",
-        "GitLabBranch" => "GitLab Branch",
-        "BitbucketURI" => "Bitbucket Plugin URI",
-        "BitbucketBranch" => "Bitbucket Branch"
-    ) );
-}
+class GitHub_Link {
 
-function GHL_plugin_link( $actions, $plugin_file, $plugin_data, $context ) {
+    /**
+     * GitHub_Link constructor.
+     */
+    private function __construct() {
+    }
 
-    // No GitHub data during search installed plugins.
-    if ( 'search' === $context )
-        return $actions;
+    /**
+     * GitHub_Link init function.
+     */
+    public static function init() {
+      $self = new self();
+      $self->init_hooks();
+    }
 
-    $link_template = '<a href="%s" title="%s" target="_blank"><img src="%s" style="width: 16px; height: 16px; vertical-align:-3px;" height="16" width="16" alt="%s" />%s</a>';
-    $wp_link_template = '<a href="%s" title="%s" target="_blank"><span style="width: 16px; height: 16px; color:#2880A8; font-size:18px; vertical-align:-3px;" class="github-link dashicons dashicons-wordpress"></span></a>';
 
-    $on_wporg = false;
-    _maybe_update_plugins();
-    $plugin_state = get_site_transient( 'update_plugins' );
-    if ( isset( $plugin_state->response[ $plugin_file ] )
-        || isset( $plugin_state->no_update[ $plugin_file ] )
-    ) {
-        $on_wporg = true;
-    }
+    function init_hooks() {
+        include_once ABSPATH . '/wp-admin/includes/plugin.php';
+
+        $this->extra_plugin_headers = [ 'GitHub Plugin URI', 'GitLab Plugin URI', 'Bitbucket Plugin URI' ];
+
+        add_filter( "extra_plugin_headers", [ &$this, "extra_plugin_headers" ] );
```

Comment:
GitHub Updater moved away from using this filter in favor of using get_plugin_data() to find and return header data. It was much less flaky overall.

Reply:
All right, after this modification I'll merge this and tag v1.0.0

Reply:
@afragen I try and let you know, however from what I remember, is it still not necessary to add additional headers?

- [get_plugin_data](https://developer.wordpress.org/reference/functions/get_plugin_data/)
- [get_file_data](https://developer.wordpress.org/reference/functions/get_file_data/)

Reply:
I have tried the following:

```php
$this->extra_plugin_headers = [ 'GitHub Plugin URI', 'GitLab Plugin URI', 'Bitbucket Plugin URI' ];

// add_filter( "extra_plugin_headers", [ &$this, "extra_plugin_headers" ] );
```

But doing so the following hooks:

- [plugin_action_links_{$plugin_file}](https://developer.wordpress.org/reference/hooks/plugin_action_links_plugin_file/)
- [network_admin_plugin_action_links_{$plugin_file}](https://developer.wordpress.org/reference/hooks/network_admin_plugin_action_links_plugin_file/)

cannot provide information about thees additional headers.

> GitHub Updater moved away from using this filter in favor of using get_plugin_data() to find and return header data. It was much less flaky overall.

Actually i don't yet understand how your plugin can read them without adding them to the list of known header..

Reply:
@afragen I think we **really** need you as I have no clue what is going on.